examples/authentication/server/src/token_exchanger.rs

```diff
@@ -0,0 +1,144 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Utilities for exchanging authorisation codes for identity tokens and validating the resulting
+//! tokens.
+//!
+//! For reference, see:
+//! https://developers.google.com/identity/protocols/oauth2/openid-connect#exchangecode
+
+use jsonwebtoken::{decode, decode_header, Algorithm, DecodingKey, Header, Validation};
+use log::info;
+use serde::{Deserialize, Serialize};
+use std::{error, fmt};
+use url::form_urlencoded;
+
+use crate::http_client;
+
+/// Claims encoded into the identity token
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Claims {
+    aud: String,
+    exp: usize,
+    iat: usize,
+    iss: String,
+    sub: String,
+    email: String,
+    // TODO(#922): Add nonce
+}
+
+#[derive(Deserialize, Debug)]
+struct TokenResponse {
+    expires_in: usize,
+    id_token: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct KeyList {
+    keys: Vec<Key>,
+}
+
+#[derive(Clone, Deserialize, Debug)]
+struct Key {
+    // The signing algorithm for the key.
+    alg: String,
+    // Base64 encoded RSA exponent.
+    e: String,
+    // The key ID.
+    kid: String,
+    // Base64 encoded RSA modulus.
+    n: String,
+}
+
+pub async fn exchange_code_for_token(
+    code: &str,
+    client_id: &str,
+    client_secret: &str,
+) -> Result<String, Box<dyn error::Error>> {
+    let body: String = form_urlencoded::Serializer::new(String::new())
+        .append_pair("code", code)
+        .append_pair("client_id", client_id)
+        .append_pair("client_secret", client_secret)
+        // TODO(#922): Pass the actual redirect URI from the client app.
+        .append_pair("redirect_uri", "http://127.0.0.1:8089")
+        .append_pair("grant_type", "authorization_code")
+        .finish();
+
+    // TODO(#923): Get the token URL from the discovery document.
+    let token_uri = "https://oauth2.googleapis.com/token".parse()?;
+    let response = http_client::post(token_uri, &body, "application/x-www-form-urlencoded").await?;
```

Comment:
Perhaps define a const for`"application/x-www-form-urlencoded"` somewhere?

Reply:
Code replaced by using `reqwest`